x/subspaces/keeper/msg_server.go

```diff
@@ -0,0 +1,499 @@
+			sdk.NewAttribute(sdk.AttributeKeySender, signer.String()),
+		),
+		sdk.NewEvent(
+			types.ActionSetUserGroupPermissions,
+			sdk.NewAttribute(types.AttributeKeySubspaceID, fmt.Sprintf("%d", msg.SubspaceID)),
+			sdk.NewAttribute(types.AttributeKeyUserGroupID, fmt.Sprintf("%d", msg.GroupID)),
+		),
+	})
+
+	return &types.MsgSetUserGroupPermissionsResponse{}, nil
+}
+
+// DeleteUserGroup defines a rpc method for MsgDeleteUserGroup
+func (k msgServer) DeleteUserGroup(goCtx context.Context, msg *types.MsgDeleteUserGroup) (*types.MsgDeleteUserGroupResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the subspace exists
+	if !k.HasSubspace(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	// Check if the group exists
+	if !k.HasUserGroup(ctx, msg.SubspaceID, msg.GroupID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "group %d could not be found", msg.GroupID)
+	}
+
+	signer, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid signer address: %s", msg.Signer)
+	}
+
+	// Check for permissions
+	if !k.HasPermission(ctx, msg.SubspaceID, signer, types.PermissionManageGroups) {
+		return nil, sdkerrors.Wrap(types.ErrPermissionDenied, "you cannot delete user groups in this subspace")
+	}
+
+	// Delete the group
+	k.Keeper.DeleteUserGroup(ctx, msg.SubspaceID, msg.GroupID)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Signer),
+		),
+		sdk.NewEvent(
+			types.EventTypeDeleteUserGroup,
+			sdk.NewAttribute(types.AttributeKeySubspaceID, fmt.Sprintf("%d", msg.SubspaceID)),
+			sdk.NewAttribute(types.AttributeKeyUserGroupID, fmt.Sprintf("%d", msg.GroupID)),
+		),
+	})
+
+	return &types.MsgDeleteUserGroupResponse{}, nil
+}
+
+// AddUserToUserGroup defines a rpc method for MsgAddUserToUserGroup
+func (k msgServer) AddUserToUserGroup(goCtx context.Context, msg *types.MsgAddUserToUserGroup) (*types.MsgAddUserToUserGroupResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the subspace exists
+	if !k.HasSubspace(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	// Check if the group exists
+	if !k.HasUserGroup(ctx, msg.SubspaceID, msg.GroupID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "group %d could not be found", msg.GroupID)
+	}
+
+	signer, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid signer address: %s", msg.Signer)
+	}
+
+	// Check the permissions
+	if !k.HasPermission(ctx, msg.SubspaceID, signer, types.PermissionSetPermissions) {
```

Comment:
Actually, the `PermissionManageGroups` is thought to allow users to edit the details of groups (the name, description, etc) and not the members or the associated permissions. Since adding/removing a user to a group is the same thing as setting their permissions individually, I think it's better to require the `PermissionManageGroups` here instead.

Reply:
🤔 Did you mean `PermissionSetPermissions`?

Reply:
Yes, thanks. Fixed